internal/gps/source_cache_bolt.go

```diff
@@ -0,0 +1,437 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gps
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/boltdb/bolt"
+	"github.com/golang/dep/internal/gps/pkgtree"
+	"github.com/pkg/errors"
+)
+
+// singleSourceCacheBolt implements a singleSourceCache backed by a persistent BoltDB file.
+// Stored values are timestamped, and the `epoch` field limits the age of returned values.
+// Database access methods are safe for concurrent use with each other (excluding close).
+//
+// Implementation:
+//
+// At the top level there are buckets for (1) versions and (2) revisions.
+//
+// 1) Versions buckets hold version keys with revision values:
+//
+//	Bucket: "versions:<timestamp>"
+//	Keys: "branch:<branch>", "defaultBranch:<branch>", "ver:<version>"
+//	Values: "<revision>"
+//
+// 2) Revision buckets hold (a) manifest and lock data for various ProjectAnalyzers,
+// (b) package trees, and (c) version lists.
+//
+//	Bucket: "rev:<revision>"
+//
+// a) Manifest and Lock info are stored in a bucket derived from ProjectAnalyzer.Info:
+//
+//	Sub-Bucket: "info:<name>.<version>:<timestamp>"
+//	Sub-Bucket: "manifest", "lock"
+//	Keys/Values: Manifest or Lock fields
+//
+// b) Package tree buckets contain package import path keys and package-or-error buckets:
+//
+//	Sub-Bucket: "ptree:<timestamp>"
+//	Sub-Bucket: "<import_path>"
+//	Key/Values: PackageOrErr fields
+//
+// c) Revision-versions buckets contain lists of version values:
+//
+//	Sub-Bucket: "versions:<timestamp>"
+//	Keys: "<sequence_number>"
+//	Values: "branch:<branch>", "defaultBranch:<branch>", "ver:<version>"
+type singleSourceCacheBolt struct {
+	ProjectRoot
+	db     *bolt.DB
+	epoch  int64       // getters will not return values older than this unix timestamp
+	logger *log.Logger // info logging
+}
+
+// newBoltCache returns a new singleSourceCacheBolt backed by a project's BoltDB file under the cache directory.
+func newBoltCache(cd string, pi ProjectIdentifier, epoch int64, logger *log.Logger) (*singleSourceCacheBolt, error) {
+	path := sourceCachePath(cd, pi.normalizedSource()) + ".db"
```

Comment:
i mean, certainly this has the benefit that simultaneous actions on different sources will not result in a bunch of competing write transactions 🤔

Reply:
I went into it assuming a single file, but given the current architecture it was cleaner/simpler to start this way, and there are definitely trade-offs to consider, so I didn't want to run ahead. My intuition is that bolt can probably juggle transactions to isolated top level buckets just fine, but I have no idea what that coordination costs are. On the other hand, reducing the number of files could be beneficial to avoid limits, or just to speed up opening/closing.

It shouldn't be too hard to convert to a single file. Most of the work/mess would be in the integration. We'll just have to store a bolt reference (or a fancier `cacheFactory` sort of type) in the `sourceCoordinator` to pass to new `newSourceGateway` calls or something like that.

Reply:
> Reducing the number of files could be beneficial to avoid limits,

ah, yes, this is a good point, particularly given the issues we've been having with it later.

i'm OK with sticking with multiple files for now, though. we can revisit before we make the code live. all we need to do now is keep an eye out for decisions that might make it harder to switch to single file.

Reply:
Well I can either work on this, or a different encoding next. Unless you want to merge something intermediate first.

Reply:
let's do this first, and we can leave the encoding for this PR. that can be a follow-up, primarily once we're happy with tests. i haven't actually dug into those, yet.